src/agent/MessageSender.ts

```diff
@@ -33,23 +34,50 @@ export class MessageSender {
     return this._outboundTransporter
   }
 
-  public async packMessage(outboundMessage: OutboundMessage): Promise<OutboundPackage> {
+  public async packMessage(outboundMessage: OutboundMessage, service: DidCommService): Promise<OutboundPackage> {
     const { connection, payload } = outboundMessage
     const { verkey, theirKey } = connection
-    const endpoint = this.transportService.findEndpoint(connection)
+    const endpoint = service.serviceEndpoint
     const message = payload.toJSON()
     this.logger.debug('outboundMessage', { verkey, theirKey, message })
+    const keys = {
+      recipientKeys: service.recipientKeys,
+      routingKeys: service.routingKeys || [],
+      senderVk: connection.verkey,
+    }
+    const wireMessage = await this.envelopeService.packMessage(keys, outboundMessage.payload)
     const responseRequested = outboundMessage.payload.hasReturnRouting()
-    const wireMessage = await this.envelopeService.packMessage(outboundMessage)
     return { connection, payload: wireMessage, endpoint, responseRequested }
   }
 
   public async sendMessage(outboundMessage: OutboundMessage): Promise<void> {
     if (!this.outboundTransporter) {
       throw new AriesFrameworkError('Agent has no outbound transporter!')
     }
-    const outboundPackage = await this.packMessage(outboundMessage)
-    outboundPackage.session = this.transportService.findSession(outboundMessage.connection.id)
-    await this.outboundTransporter.sendMessage(outboundPackage)
+
+    const services = this.transportService.findServices(outboundMessage.connection)
+    if (services.length === 0) {
+      throw new AriesFrameworkError(`Connection with id ${outboundMessage.connection.id} has no service!`)
+    }
+
+    let success = false
+    for await (const service of services) {
+      this.logger.debug(`Sending outbound message to service:`, { service })
+      if (!success) {
+        try {
+          const outboundPackage = await this.packMessage(outboundMessage, service)
+          outboundPackage.session = this.transportService.findSession(outboundMessage.connection.id)
+          await this.outboundTransporter.sendMessage(outboundPackage)
+          success = true
+        } catch (error) {
+          this.logger.debug(
+            `Sending outbound message to service with id ${service.id} failed with the following error:`,
+            {
+              error,
+            }
+          )
+        }
+      }
+    }
```

Comment:
```suggestion
    for (const service of services) {
      this.logger.debug(`Trying to send outbound message to service:`, { service })
        try {
          const outboundPackage = await this.packMessage(outboundMessage, service)
          outboundPackage.session = this.transportService.findSession(outboundMessage.connection.id)
          await this.outboundTransporter.sendMessage(outboundPackage)
          
          break
          // Or return
        } catch (error) {
          this.logger.debug(
            `Sending outbound message to service with id ${service.id} failed with the following error:`,
            {
              error,
            }
          )
      }
    }
```

Reply:
Not a fan of the `success` property

Reply:
I also don't feel good about the `success` variable, but I don't consider the `break` statement as a much better solution. It doesn't tell you the what's the intent of breaking the loop. However, I see a benefit in removing the condition and variable so I can make the change.